File: runme/mcgen/local/datan/solsrc/s1tim.c
#include <stdio.h>
#include <math.h>
#include <string.h>
#include "datsrc.h"
#include "grsrc.h"

/* Table of constant values */

static integer c__1 = 1, c__2 = 2, c__3 = 3, c__4 = 4, c__75 = 75;
static double c_dm20 = -20., c_d0 = 0., c_d1 = 1., c_dp2 = .2, c_dp8 = .8, 
              c_dmp314 = -.314, c_d1p1 = 1.1;

struct {
    double ata1at[861] /* was [21][41] */, etadel[440], coneta[440],
           t[440], y[400], scrat[441] /* was [21][21] */,
           a[861] /* was [41][21] */, ata1[441] /* was [21][21] */, eta[440];
} LIBDATA s1tcom;

main(void)
{
    /* Local variables */
    static integer i, k, l, m, n, nws;
    static double p, x, sigma, time, ymin, ymax, d, f;
    static char capt[75], string[1];
    /* System generated locals */
    integer i__1;
    double d__1;
/* identify program to user */
    printf("%s\n", "Program S1TIM simulates data with errors");
    printf("%s\n", "performs time series analysis on them");
    printf("%s\n\n", "and presents results graphically.");
/* ask for input */
    printf("%s\n", "Enter power m (>=0):");
    printf("%s", "> ");
    scanf("%li", &m);
    printf("%s\n", "Enter SIGMA (>0.):");
    printf("%s", "> ");
    scanf("%lf", &sigma);
    printf("%s\n", "Enter k ( 0 < k < 21 ):");
    printf("%s", "> ");
    scanf("%li", &k);
    printf("%s\n", "Enter l ( 0 < l < k ):");
    printf("%s", "> ");
    scanf("%li", &l);
    printf("%s\n", "Enter P ( 0 < P < 1.0 ):");
    printf("%s", "> ");
    scanf("%lf", &p);
/* ask for number of workstation */
    grnbws_();
    printf("%s\n", "Please, enter number of workstation:");
    printf("%s", "> ");
    scanf("%li", &nws);
    (void)getchar();
/* generate time series */
    n = 200;
    rnstnr_(s1tcom.y, &n);
    i__1 = n;
    for (i = 1; i <= i__1; ++i) {
	s1tcom.y[i - 1] *= sigma;
	time = (double) i;
	x = (time - 100.) * .02;
	f = pow(x, (double)m);
	s1tcom.y[i - 1] += f;
    }
    timser_(s1tcom.y, &n, &k, &l, &p, s1tcom.eta, s1tcom.coneta, s1tcom.a,
           s1tcom.ata1, s1tcom.ata1at, s1tcom.scrat);
    gropen_();
    gropws_(&nws);
    ymin = s1tcom.y[0];
    ymax = s1tcom.y[0];
    i__1 = n;
    for (i = 2; i <= i__1; ++i) {
/* Computing MIN */
	d__1 = s1tcom.y[i - 1];
	ymin = min(d__1,ymin);
/* Computing MAX */
	d__1 = s1tcom.y[i - 1];
	ymax = max(d__1,ymax);
    }
    d = ymax - ymin;
    ymin -= d * .2;
    ymax += d * .2;
    d__1 = (double) (n + 20);
    grwncc_(&c_dm20, &d__1, &ymin, &ymax);
    grvwwc_(&c_d0, &c_d1, &c_dp2, &c_dp8);
    grwnwc_(&c_dmp314, &c_d1p1, &c_d0, &c_d1);
    grboun_();
    grfram_();
/* write caption */
    memset(capt, ' ', 75);
    sprintf(capt, "N = %3li, k = %3li, l = %3li, P = %5.3lf", n, k, l, p);
    capt[36] = ' ';
    grtxtc_(&c_d1, capt, &c__75);
    *(unsigned char *)string = 't';
    grsclx_(string, &c__1);
    *(unsigned char *)string = 'y';
    grscly_(string, &c__1);
/* draw data points */
    grstcl_(&c__2);
    i__1 = n;
    for (i = 1; i <= i__1; ++i) {
	d__1 = (double) i;
	grdatp_(&c__1, &c_dp2, &d__1, &s1tcom.y[i - 1], &c_d0, &c_d0, &c_d0);
    }
/* draw moving averages */
    grstcl_(&c__3);
    i__1 = n + (k << 1);
    for (i = 1; i <= i__1; ++i) {
	s1tcom.t[i - 1] = (double) (i - k);
    }
    i__1 = n + (k << 1);
    grplin_(&i__1, s1tcom.t, s1tcom.eta);
    grstcl_(&c__4);
/* draw confidence limits */
    i__1 = n + (k << 1);
    for (i = 1; i <= i__1; ++i) {
	s1tcom.etadel[i - 1] = s1tcom.eta[i - 1] + s1tcom.coneta[i - 1];
    }
    i__1 = n + (k << 1);
    grplin_(&i__1, s1tcom.t, s1tcom.etadel);
    i__1 = n + (k << 1);
    for (i = 1; i <= i__1; ++i) {
	s1tcom.etadel[i - 1] = s1tcom.eta[i - 1] - s1tcom.coneta[i - 1];
    }
    i__1 = n + (k << 1);
    grplin_(&i__1, s1tcom.t, s1tcom.etadel);
    grclse_();
    return 0;
} /* main */

File: runme/mcgen/local/datan/solsrc/s1tim.f
      PROGRAM S1TIM
      IMPLICIT DOUBLE PRECISION(A-H,O-Z)
      PARAMETER(MAXN=400,MAXL=20,MAXK=20)
      DIMENSION Y(MAXN),T(MAXN+2*MAXK),ETA(MAXN+2*MAXK)
      DIMENSION ETADEL(MAXN+2*MAXK),CONETA(MAXN+2*MAXK)
      DIMENSION ATA1(MAXL+1,MAXL+1),ATA1AT(MAXL+1,2*MAXK+1)
      DIMENSION SCRAT(MAXL+1,MAXL+1),A(2*MAXK+1,MAXL+1)
      COMMON /CS1TIM/ ATA1,AT1AT,A,SCRAT
      CHARACTER*1 STRING
      CHARACTER*75 CAPT
C identify program to user
      WRITE(*,*)' Program S1TIM simulates data with errors'
      WRITE(*,*)' performs time series analysis on them'
      WRITE(*,*)' and presents results graphically'
      WRITE(*,*)' '
C ask for input
      WRITE(*,*)'  Enter power m (>=0)'
      WRITE(*,*)' >'
      READ(*,*)M
      WRITE(*,*)' Enter SIGMA (>0.)'
      WRITE(*,*)' >'
      READ(*,*)SIGMA
      WRITE(*,*)' Enter k ( 0 < k < 21 )'
      WRITE(*,*)' >'
      READ(*,*)K
      WRITE(*,*)' Enter l ( 0 < l < k )'
      WRITE(*,*)' >'
      READ(*,*)L
      WRITE(*,*)' Enter P ( 0 < P < 1.0 )'
      WRITE(*,*)' >'
      READ(*,*)P
C ask for number of workstation
      CALL GRNBWS()
      WRITE(*,*)' Please, enter number of workstation:'
      WRITE(*,*)' >'
      READ(*,*)NWS
C generate time series
      N=200
      CALL RNSTNR(Y,N)
      DO 1 I=1,N
        Y(I)=Y(I)*SIGMA
        TIME=DBLE(I)
        X=0.02D0*(TIME-100.D0)
        F=X**M
        Y(I)=Y(I)+F
1     CONTINUE
      CALL TIMSER(Y,N,K,L,P,ETA,CONETA,A,ATA1,ATA1AT,SCRAT)
      CALL GROPEN
      CALL GROPWS(NWS)
      YMIN=Y(1)
      YMAX=Y(1)
      DO 7 I=2,N
        YMIN=MIN(Y(I),YMIN)
        YMAX=MAX(Y(I),YMAX)
7     CONTINUE
      D=YMAX-YMIN
      YMIN=YMIN-0.2D0*D
      YMAX=YMAX+0.2D0*D
      CALL GRWNCC(-20.D0,DBLE(N+20),YMIN,YMAX)
      CALL GRVWWC(0.D0,1.D0,.2D0,.8D0)
      CALL GRWNWC(-.314D0,1.1D0,0.D0,1.D0)
      CALL GRBOUN
      CALL GRFRAM
C write caption
      CAPT='N = ***, k = ***, l = ***, P = *****'
      WRITE(CAPT(5:7),'(I3)')N
      WRITE(CAPT(14:16),'(I3)')K
      WRITE(CAPT(23:25),'(I3)')L
      WRITE(CAPT(32:36),'(F5.3)')P
      CALL GRTXTC(1.D0,CAPT,75)
      STRING='t'
      CALL GRSCLX(STRING,1)
      STRING='y'
      CALL GRSCLY(STRING,1)
C draw data points
      CALL GRSTCL(2)
      DO 10 I=1,N
        CALL GRDATP(1,.2D0,DBLE(I),Y(I),0.D0,0.D0,0.D0)
10    CONTINUE
C draw moving averages
      CALL GRSTCL(3)
      DO 20 I=1,N+2*K
        T(I)=DBLE(I-K)
20    CONTINUE
      CALL GRPLIN(N+2*K,T,ETA)
      CALL GRSTCL(4)
C draw confidence limits
      DO 30 I=1,N+2*K
        ETADEL(I)=ETA(I)+CONETA(I)
30    CONTINUE
      CALL GRPLIN(N+2*K,T,ETADEL)
      DO 40 I=1,N+2*K
        ETADEL(I)=ETA(I)-CONETA(I)
40    CONTINUE
      CALL GRPLIN(N+2*K,T,ETADEL)
      CALL GRCLSE
      END
